planner/core/common_plans.go

```diff
@@ -261,25 +261,15 @@ func (e *Execute) OptimizePreparedPlan(ctx context.Context, sctx sessionctx.Cont
 			vars.PreparedParams = append(vars.PreparedParams, val)
 		}
 	}
-	e.TxnScope = oracle.GlobalTxnScope
-	var snapshotTS uint64
-	if preparedObj.SnapshotTSEvaluator != nil {
-		if vars.InTxn() {
-			return ErrAsOf.FastGenWithCause("as of timestamp can't be set in transaction.")
-		}
-		// if preparedObj.SnapshotTSEvaluator != nil, it is a stale read SQL:
-		// which means its infoschema is specified by the SQL, not the current/latest infoschema
-		var err error
-		snapshotTS, err = preparedObj.SnapshotTSEvaluator(sctx)
-		if err != nil {
-			return errors.Trace(err)
-		}
+	snapshotTS, txnScope, isStaleness, err := e.handleExecuteBuilderOption(sctx, preparedObj)
+	if err != nil {
+		return err
+	}
+	if snapshotTS != 0 && isStaleness {
```

Comment:
I think we can just use `isStaleness`. `snapshotTS != 0` is the implicit way before there is `isStaleness` flag.